lib/PersonalityProvider.jsm

```diff
@@ -0,0 +1,249 @@
+          continue;
+        }
+        if (model.data.model_type === "nb") {
+          nbTaggers.push(new NaiveBayesTextTagger(model.data));
+        } else if (model.data.model_type === "nmf") {
+          nmfTaggers[model.data.parent_tag] = new NmfTextTagger(model.data);
+        }
+      }
+      this.dispatch(ac.PerfEvent({
+        event: "PERSONALIZATION_V2_TAGGERS_DURATION",
+        value: Math.round(perfService.absNow() - startTaggers),
+      }));
+      this.taggers = {nbTaggers, nmfTaggers};
+    }
+    const startRecipeExecutor = perfService.absNow();
+    const recipeExecutor = new RecipeExecutor(this.taggers.nbTaggers, this.taggers.nmfTaggers);
+    this.dispatch(ac.PerfEvent({
+      event: "PERSONALIZATION_V2_RECIPE_EXECUTOR_DURATION",
+      value: Math.round(perfService.absNow() - startRecipeExecutor),
+    }));
+    return recipeExecutor;
+  }
+
+  /**
+   * Grabs a slice of browse history for building a interest vector
+   */
+  async fetchHistory(columns, beginTimeSecs, endTimeSecs) {
+    let sql = `SELECT url, title, visit_count, frecency, last_visit_date, description
+    FROM moz_places
+    WHERE last_visit_date >= ${beginTimeSecs * 1000000}
+    AND last_visit_date < ${endTimeSecs * 1000000}`;
+    columns.forEach(requiredColumn => {
+      sql += ` AND IFNULL(${requiredColumn}, "") <> ""`;
+    });
+    sql += " LIMIT 30000";
+
+    const {activityStreamProvider} = NewTabUtils;
+    const history = await activityStreamProvider.executePlacesQuery(sql, {
+      columns,
+      params: {},
+    });
+
+    return history;
+  }
+
+  /**
+   * Examines the user's browse history and returns an interest vector that
+   * describes the topics the user frequently browses.
+   */
+  async createInterestVector() {
+    const start = perfService.absNow();
+    let interestVector = {};
+    let endTimeSecs = ((new Date()).getTime() / 1000);
+    let beginTimeSecs = endTimeSecs - this.interestConfig.history_limit_secs;
+    let history = await this.fetchHistory(this.interestConfig.history_required_fields, beginTimeSecs, endTimeSecs);
+
+    this.dispatch(ac.PerfEvent({
+      event: "PERSONALIZATION_V2_HISTORY_SIZE",
+      value: history.length,
+    }));
+
+    for (let historyRec of history) {
+      let ivItem = this.recipeExecutor.executeRecipe(
+        historyRec,
+        this.interestConfig.history_item_builder);
+      if (ivItem === null) {
+        continue;
+      }
+      interestVector = this.recipeExecutor.executeCombinerRecipe(
+        interestVector,
+        ivItem,
+        this.interestConfig.interest_combiner);
+      if (interestVector === null) {
+        return null;
+      }
+    }
+
+    const finalResult = this.recipeExecutor.executeRecipe(
+      interestVector,
+      this.interestConfig.interest_finalizer);
+
+    this.dispatch(ac.PerfEvent({
+      event: "PERSONALIZATION_V2_CREATE_INTEREST_VECTOR_DURATION",
+      value: Math.round(perfService.absNow() - start),
```

Comment:
You start the timer before `fetchHistory`, is that intended? If not, let's move it down a bit. Otherwise, ignore this comment.

Reply:
Yeah it was my initial intent to track the time the whole function takes, but I'm open to tracking the duration of just the calculation. Really I didn't have a strong opinion on it, so either or is fine with me.